File: communication.lua
local api = require("api")
local json = require("cjson")
local tools = require("tools")

local _G = _G

do
    local registered = {}
    AskProceed = function(tag)
        if registered[tag] then
            return true
        end
        print("Proceed?[(Y)es/(n)o/(a)bort/yesforall]")
        local input = io.read()
        if input == "Y" or input == "y" then
            return true
        elseif input == "n" or input == "N" then
            return false
        elseif input == "a" or input == "A" then
            print("Abort.")
            os.exit(0)
        elseif input == "yesforall" then
            registered[tag] = true
            return true
        else
            print("Invalid input. Please enter 'Y', 'n', 'a', or 'yesforall'.")
            return AskProceed(tag)
        end
    end
end

local communication = {}

local registered = {}

communication.__index = communication

function communication.onTool(openai_api, id, tool_name, arguments)
    local tool = tools[tool_name]
    if not tool then
        openai_api:toolcall(id, tool_name, "Tool not found")
        return
    end
    if tool.action then
        local result = tool.action(arguments)
        openai_api:toolcall(id, tool_name, result.content)
    end
end

function communication:new(model, role_name)
    for _, registered_communication in ipairs(registered) do
        if registered_communication.role_name == role_name then
            registered_communication.model = model
            registered_communication.model_name = model.name
            return registered_communication
        end
    end
    local mcommunication = {}
    setmetatable(mcommunication, self)

    mcommunication.register = true
    mcommunication.api = api[model.api_type].create(model, role_name)
    mcommunication.api.onTool = communication.onTool
    mcommunication.model_name = model.name
    mcommunication.model = model
    mcommunication.id = os.time()

    table.insert(registered, mcommunication)
    return mcommunication
end

function communication:call(tool_name, arguments)
    local tools = self.model.tools
    for _, tool in ipairs(tools) do
        if tool.name == tool_name then
            if tool.action then
                local result = tool.action(arguments)
                return true, result
            end
        end
    end
    return false, nil
end

function communication:appendUserMessage(message)
    if type(self) ~= "table" or not self.register then
        error("self: recognize failed")
    end
    self.api:appendUserMessage(message)
end

function communication:send()
    if type(self) ~= "table" or not self.register then
        error("self: recognize failed")
    end
    self.api:send()
    return message
end

return communication

File: communication.py
"""Communication manager — orchestrates conversations, tool dispatch, and role delegation."""

from __future__ import annotations

import os
import time

from api import OpenAIAPI
from models import MODELS


# --- AskProceed ---
_approved: set[str] = set()
_auto_approve = os.getenv("AI_TEAM_AUTO_APPROVE", "").lower() in ("1", "true", "yes")


def ask_proceed(tag: str) -> bool:
    """Ask the user whether to proceed with a tool action."""
    if tag in _approved or _auto_approve:
        return True
    while True:
        resp = input(f"Proceed with {tag}? [(Y)es/(n)o/(a)bort/yesforall]: ").strip()
        if resp in ("Y", "y", ""):
            return True
        elif resp in ("n", "N"):
            return False
        elif resp in ("a", "A"):
            print("Abort.")
            raise SystemExit(0)
        elif resp == "yesforall":
            _approved.add(tag)
            return True
        else:
            print("Invalid input. Enter Y, n, a, or yesforall.")


# --- Communication ---
_registered: list["Communication"] = []


class Communication:
    """Manages a conversation with a specific role, backed by an API client."""

    def __init__(self, model_key: str, role_name: str):
        self.role_name = role_name
        self.model_key = model_key
        self.model = MODELS[model_key]
        self.api = OpenAIAPI(model_key, role_name)
        self.id = int(time.time())

    def append_user_message(self, message: str):
        self.api.append_user_message(message)

    def send(self):
        self.api.send()

    @property
    def last_response(self) -> str | None:
        """Get the last assistant message content."""
        for msg in reversed(self.api.messages):
            if msg.get("role") == "assistant" and msg.get("content"):
                return msg["content"]
        return None

    @staticmethod
    def get_or_create(model_key: str, role_name: str) -> "Communication":
        """Reuse an existing Communication for the same role, or create a new one."""
        for comm in _registered:
            if comm.role_name == role_name:
                comm.model_key = model_key
                comm.model = MODELS[model_key]
                return comm
        comm = Communication(model_key, role_name)
        _registered.append(comm)
        return comm
